v2/internal/reconcilers/mysql/aad_user.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package mysql
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+
+	asomysql "github.com/Azure/azure-service-operator/v2/api/dbformysql/v1"
+	"github.com/Azure/azure-service-operator/v2/internal/identity"
+	. "github.com/Azure/azure-service-operator/v2/internal/logging"
+	"github.com/Azure/azure-service-operator/v2/internal/resolver"
+	mysqlutil "github.com/Azure/azure-service-operator/v2/internal/util/mysql"
+)
+
+const Scope = "https://ossrdbms-aad.database.windows.net/.default"
+
+type aadUser struct {
+	user               *asomysql.User
+	resourceResolver   *resolver.Resolver
+	credentialProvider identity.CredentialProvider
+	log                logr.Logger
+}
+
+var _ Connector = &aadUser{}
+
+func (u *aadUser) CreateOrUpdate(ctx context.Context) error {
+	db, err := u.connectToDB(ctx)
+	if err != nil {
+		return err
+	}
+	defer db.Close()
+
+	u.log.V(Status).Info("Creating MySQL AAD user")
+
+	username := u.username()
+	err = mysqlutil.CreateOrUpdateAADUser(ctx, db, u.user.Spec.AzureName, username)
+	if err != nil {
+		return errors.Wrap(err, "failed to create user")
+	}
+
+	// Ensure that the privileges are set
+	err = mysqlutil.ReconcileUserServerPrivileges(ctx, db, username, u.user.Spec.Hostname, u.user.Spec.Privileges)
+	if err != nil {
+		return errors.Wrap(err, "ensuring server roles")
+	}
+
+	err = mysqlutil.ReconcileUserDatabasePrivileges(ctx, db, username, u.user.Spec.Hostname, u.user.Spec.DatabasePrivileges)
+	if err != nil {
+		return errors.Wrap(err, "ensuring database roles")
+	}
+
+	u.log.V(Status).Info("Successfully reconciled MySQLUser")
+	return nil
+}
+
+func (u *aadUser) Delete(ctx context.Context) error {
+	db, err := u.connectToDB(ctx)
+	if err != nil {
+		return err
+	}
+	defer db.Close()
+
+	// TODO: There's still probably some ways that this user can be deleted but that we don't detect (and
+	// TODO: so might cause an error triggering the resource to get stuck).
+	// TODO: Cases where the server is in the process of being
+	// TODO: deleted (or all system tables have been wiped?) might also exist...
```

Comment:
Does this need to be addressed before merging?

Reply:
No, we had this comment even on the existing MySQL local user support. This is just saying that, we're not confident that the deletion logic here is perfect. Unfortunately actually _finding_ the edge cases where things get stuck is very tricky. It passes in all of our tests, but could get stuck for some time during something like a SQL failover.

I think we just need to keep our eyes open for user reports that suggest an issue here and fix them as they arise.